framework/sandstone.cpp

```diff
@@ -2611,70 +2610,93 @@ static int exec_mode_run(int argc, char **argv)
 // Triage run attempts to figure out which socket(s) are causing test failures.
 // It simply removes sockets from the cpu_set_t one by one until the failures no
 // longer observed.
-// Returns the list of faulty sockets. Memory is allocated, caller must free.
-// TODO: Current implementation only returns 1 socket. However, the signature is
-// generic vector<int> for the future improvements.
+// Returns the list with "suspected" faulty sockets
 static vector<int> run_triage(vector<const struct test *> &triage_tests, const LogicalProcessorSet &enabled_cpus)
 {
     using Socket = Topology::Package;
 
     int orig_verbosity;
     Topology topo = Topology::topology();
-    vector<Socket>::iterator it;
-    vector<int> disabled_sockets;
     vector<int> result; // faulty sockets
     int ret = EXIT_SUCCESS;
     bool ever_failed = false;
 
-    it = topo.packages.begin();
-    if (topo.packages.empty())
-        return result;                  // shouldn't happen!
-
-    if (topo.packages.size() == 1) {
-        result.push_back(it->id);
-        return result;
-    }
-
     // backup the original verbosity
     orig_verbosity = sApp->verbosity;
     sApp->verbosity = -1;
 
-    for (; it != topo.packages.end(); disabled_sockets.push_back(it->id), ++it) {
-        vector<Socket>::iterator eit = it;
+    // first "non-empty" package found in the topology
+    vector<Socket>::iterator first_it = topo.packages.end();
+    // first "non-empty" package being "removed" from the list
+    vector<Socket>::iterator removed_it = topo.packages.end();
+    // last "non-empty" package being added
+    vector<Socket>::iterator last_it = topo.packages.end();
+
+    vector<Socket>::iterator it = topo.packages.begin();
+    while (it != topo.packages.end()) {
+        vector<Socket>::iterator eit;
         LogicalProcessorSet run_cpus = {};
-        int k = 0;
 
-        for (; eit != topo.packages.end(); ++eit)
-            run_cpus.add_package(*eit);
+        // merge all CPUs from all packages in the list (w/ respect to the removed packages)
+        for (eit = it; eit != topo.packages.end(); ++eit) {
+            if (!(*eit).cores.empty()) {
+                if (first_it == topo.packages.end()) {
+                    first_it = eit;
+                }
+                last_it = eit;
+                run_cpus.add_package(*eit);
+            }
+        }
 
-        sApp->enabled_cpus = run_cpus;
-        do {
-            ret = run_tests_on_cpu_set(triage_tests, run_cpus);
-        } while (!ret && ++k < sApp->retest_count);
+        if (!run_cpus.empty()) {
```

Comment:
Would it make sense to make this:
```c++
    if (run.cpus.empty())
        continue;
```
?

That would mean not running the code at 'find first non-empty package to "remove"'.

Reply:
'break' instead of 'continue' will do (as next packages will have empty lists as well). Decided to have more "linear", but I'm open for the change.